cpu/arm7_common: align stacks correctly

maribu opened the pull request:
### Contribution description

`thread_stack_init()` didn't correctly set up the stack alignment. This fixes the issue and brings the function closer to the Cortex M version, laying the groundwork for future code duplication.

### Testing procedure

#### Via `tests/unittests`

```
$ make BOARD=msba2 tests-printf_float -C tests/unittests flash test
```

In `master`, this fails. With this PR, it succeeds.

#### Via `tests/thread_stack_alignment`

The new test application `tests/thread_stack_alignment` provides a bit more insight in the actual issue.

##### `master`

```
make BOARD=msba2 flash test -C tests/thread_stack_alignment
```

The test fails for four out of eight possible alignments.

<details><summary>Full log output</summary>

```
Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2022.10-devel-880-g3a9f3-tmp)
Testing with a stack sized 1664 and an alignment up to 128
Testing for alignment 0: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 856 }]}
Testing for alignment 1: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 860 }]}
Testing for alignment 2: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 3: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 4: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 768 }]}
Testing for alignment 5: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 6: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 7: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 8: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 9: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 10: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 11: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 12: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 13: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 14: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 15: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 16: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 17: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 18: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 19: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 20: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 21: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 22: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 23: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 24: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 25: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 26: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 27: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 28: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 29: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 30: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 31: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 32: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 33: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 34: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 35: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 36: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 37: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 38: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 39: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 40: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 41: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 42: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 43: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 44: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 45: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 46: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 47: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 48: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 49: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 50: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 51: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 52: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 53: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 54: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 55: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 56: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 57: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 58: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 59: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 60: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 61: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 62: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 63: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 64: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 65: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 66: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 67: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 68: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 69: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 70: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 71: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 72: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 73: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 74: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 75: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 76: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 77: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 78: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 79: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 80: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 81: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 82: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 83: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 84: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 85: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 86: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 87: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 88: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 89: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 90: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 91: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 92: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 93: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 94: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 95: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 96: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 97: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 98: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 99: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 100: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 101: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 102: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 103: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 104: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 105: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 106: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 107: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 108: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 109: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 110: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 111: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 112: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 113: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 114: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 115: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 116: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 117: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 118: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 119: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 120: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 780 }]}
Testing for alignment 121: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 122: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 123: FAILED
Got "-1068331161420", expected "3.14159"
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 784 }]}
Testing for alignment 124: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 125: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 126: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 127: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
TEST FAILED
{ "threads": [{ "name": "idle", "stack_size": 160, "stack_used": 160 }]}
{ "threads": [{ "name": "main", "stack_size": 1536, "stack_used": 440 }]}
Timeout in expect script at "child.expect_exact(f"Testing for alignment {i}: OK")" (tests/thread_stack_alignment/tests/01-run.py:22)

make: *** [/home/maribu/Repos/software/RIOT/makefiles/tests/tests.inc.mk:22: test] Error 1
```

</details>


##### This PR

```
make BOARD=msba2 flash test -C tests/thread_stack_alignment
```

The test succeeds for every possible stack alignment.

<details><summary>Full log output</summary>

```
Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2022.10-devel-875-gec86d-cpu/arm7_common/alignment)
Testing with a stack sized 1664 and an alignment up to 128
Testing for alignment 0: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 804 }]}
Testing for alignment 1: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 2: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 3: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 4: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 5: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 6: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 7: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 8: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 9: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 10: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 11: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 12: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 13: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 14: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 15: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 16: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 17: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 18: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 19: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 20: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 21: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 22: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 23: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 24: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 25: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 26: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 27: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 28: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 29: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 30: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 31: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 32: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 33: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 34: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 35: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 36: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 37: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 38: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 39: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 40: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 41: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 42: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 43: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 44: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 45: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 46: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 47: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 48: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 49: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 50: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 51: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 52: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 53: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 54: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 55: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 56: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 57: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 58: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 59: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 60: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 61: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 62: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 63: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 64: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 65: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 66: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 67: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 68: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 69: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 70: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 71: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 72: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 73: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 74: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 75: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 76: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 77: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 78: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 79: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 80: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 81: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 82: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 83: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 84: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 85: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 86: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 87: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 88: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 89: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 90: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 91: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 92: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 93: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 94: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 95: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 96: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 97: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 98: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 99: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 100: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 101: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 102: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 103: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 104: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 105: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 106: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 107: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 108: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 109: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 110: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 111: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 112: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 113: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 114: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 115: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 116: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 117: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 118: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 119: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 120: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 121: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 122: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 123: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 732 }]}
Testing for alignment 124: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 724 }]}
Testing for alignment 125: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 126: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
Testing for alignment 127: OK
{ "threads": [{ "name": "test", "stack_size": 1664, "stack_used": 728 }]}
TEST PASSED
NOTE: Up to 80 B of RAM is lost when thread stacks are not properly aligned
```

### Issues/PRs references

Fixes https://github.com/RIOT-OS/RIOT/issues/11885

maribu commented:
The test application crashes on my Nucleo-F767ZI in the call to `snprintf()` for the second iteration :-/

maribu commented:
Should be ready for review. After having confirmed that the test application indeed shows a bug in the Cortex-M port rather than a bug in the test app itself, this should be fine for upstreaming :)

The two discussed approaches on fixing Cortex-M have been dropped from this PR, but can still be seen as:

- https://github.com/maribu/RIOT/tree/cpu/cortexm_common/fix_fpu_v1
- https://github.com/maribu/RIOT/tree/cpu/cortexm_common/fix_fpu_v2

benpicco commented:
Hm shouldn't we merge the fix for Cortex-M first then, as otherwise "Run tests" would always fail

maribu commented:
> as otherwise "Run tests" would always fail

IMO that is an improvement ;) Right now we have a bug in Cortex M and nothing that can detect it. With the test merged, we still have the bug in Cortex M, but we can detect it.

aabadie commented:
> otherwise "Run tests" would always fail

Same for the master nightly build which forces "Run tests" unless the fix is merged right after this test.

maribu commented:
Come on, guys! Not merging a test that is exposing a bug that *is present* upstream just to have tests passing is poor practice. The very purpose of those tests is to detect flaws. Not merging tests is defeating the very purpose of those tests.

With the same reasoning we could open a PR that removes all tests and be sure the we never see CI failure anymore.

benpicco commented:
I'd be more comfortable if there were already PRs for the fixes :wink: 

I want to avoid a situation where we just get used to some tests failing and ignore possible new failures (I'm afraid we might already be there)

aabadie commented:
> Not merging a test that is exposing a bug that _is present_ upstream just to have tests passing is poor practice

Just merging the test without the fix sounds weird to me. Ideally the test and the fix should come in the same PR.

benpicco commented:
My preference would be 7d9acc07ea794c062dee85baf20f7e70e2208ffa unless someone can convince me that floats are useful inside ISR context

maribu commented:
> Ideally the test and the fix should come in the same PR.

This PR contains a fix and a corresponding test. Only the test catches bugs unrelated to what the test was meant to catch as well.

> Just merging the test without the fix sounds weird to me.

If we are at it, why not also stop accepting bug reports / issues without a fix provided?

Seriously, blocking tests from getting upstream to create a false impression of "everything is good" is a textbook example of bad error culture.

> I'd be more comfortable if there were already PRs for the fixes

I can open a PR with the disabling of FPU state saving in IRQs. I have the feeling that backing up the FPU state on IRQ entry is something that cannot universally assumed anyway. And I bet that there is a performance penalty (even with the delayed back up) that we may not want to pay, as using the FPU in IRQs really seems to be a non-use-case.

maribu commented:
Closing in favor of https://github.com/RIOT-OS/RIOT/pull/18697

maribu commented:
Wrong PR

benpicco commented:
The new test needs a `Makefile.ci`, then this should be good

maribu commented:
Note that tests currently still fail on ESP. I think that should be relatively easy to fix, though.

maribu commented:
The test should now pass on all connected boards (hopefully!)

benpicco commented on tests/thread_stack_alignment/tests/01-run.py:
```diff
@@ -0,0 +1,39 @@
+import json
+from testrunner import run
+
+def testfunc(child):
```
static tests is unhappy

```suggestion

def testfunc(child):
```

maribu replied:
Fixed, thanks. I also added the `README.md` for the test that I apparently forgot to write originally.

maribu commented:
Many test failures on ESP. I suspect that the addition of the stack consumption printing causes issues.

I hope I can find some time to investigate soonish

benpicco commented:
We could merge 9374afa81b46d9afb6e2474b44f8950869af0c4f separately & the test and fix for esp32 later.

maribu commented:
I dropped the commit that implements `test_utils_print_stack_usage` for the custom ESP `sched_task_exit()`. I also added `FEATURES_BLACKLIST += arch_esp` to the test application for now with a comment. This way we do not need to delay the merging of the test app.
